JDK-8303950: [macos]Translucent Windows Flicker on Repaint

mickleness opened the pull request:
I'm confident about the new test case for this ticket, but the resolution is more invasive than I'd like. (It works, though.)

In short:
This introduces a new RenderingHint (in SunHints) to bypass the call in Window to `gg2d.fillRect(0, 0, getWidth(), getHeight());`

I left more detailed notes here about the proposed resolution:
https://github.com/openjdk/jdk/commit/1991fdac5dbf76ddaf73cc78a9f7c38370c9674c

I'm open to suggestions if anyone has a more elegant proposal to prevent the monitor from refreshing too soon?

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8303950](https://bugs.openjdk.org/browse/JDK-8303950): [macos]Translucent Windows Flicker on Repaint


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/12993/head:pull/12993` \
`$ git checkout pull/12993`

Update a local copy of the PR: \
`$ git checkout pull/12993` \
`$ git pull https://git.openjdk.org/jdk.git pull/12993/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12993`

View PR using the GUI difftool: \
`$ git pr show -t 12993`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12993.diff">https://git.openjdk.org/jdk/pull/12993.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/12993#issuecomment-1465257059)

mrserb commented:
It is probably better to  integrate this "clear" operation into the peer and/or repaintmanager, and exclude it in this method.

mickleness commented:
> It is probably better to integrate this "clear" operation into the peer and/or repaintmanager, and exclude it in this method.

Are you sure? (Or alternatively: maybe I don't understand what you mean?)

By "clear operation" I assume you're referring to the call in Window.java `gg2dfillRect`. This clear logic has existed since 2009, so I'm reluctant to remove it. (And if we remove it: we can just remove the entire `paint(Graphics)` method, because that's the only reason it's overridden, right?)

Here's a screenshot of the code in question & a little git annotation:
<img width="1316" alt="image" src="https://user-images.githubusercontent.com/7669569/224851605-e697e698-c4ee-439e-9e04-6b41659a12f4.png">

If you still want me to try to migrate this logic: can you give me an example of which classes you'd recommend trying to move it to?

mrserb commented:
> If you still want me to try to migrate this logic: can you give me an example of which classes you'd recommend trying to move it to?

That logic exists there to clear the window if it does not opaque and have some translucent pixels, and it was added there to clear the old content of the window before the new rendering is started. So we cannot just skip that part, but probably we can move this step closer to the place we actually start to render.

I would like to clarify two question:
 - I think most of the opaque/non-transparent components fill the background by the opaque color(including windows/frames/etc), why the problem is not reproduced in that case?
 - Why the double-buffer in the RepaintManager does not handle this? If no buffers are used, then probably we should have one to render everything to it and then blit to Window?

mickleness commented:
> I would like to clarify two question:
> 
>  1. I think most of the opaque/non-transparent components fill the background by the opaque color(including windows/frames/etc), why the problem is not reproduced in that case?

Great question. (I wasn't sure, so I had to explore this.)

The RepaintManager interprets a repaint request for any given JComponent and finds its nearest opaque ancestor to repaint.

I modified my test so the JLabel is opaque and added a debugger breakpoint in RepaintManager.paintDirtyRegions. RepaintManager enters this block of code:
```
                        if (dirtyComponent instanceof JComponent) {
                            ((JComponent)dirtyComponent).paintImmediately(
                                rect.x,rect.y,rect.width, rect.height);
                        }
                        else if (dirtyComponent.isShowing()) {
                            Graphics g = JComponent.safelyGetGraphics(
                                    dirtyComponent, dirtyComponent);
                            // If the Graphics goes away, it means someone disposed of
                            // the window, don't do anything.
                            if (g != null) {
                                g.setClip(rect.x, rect.y, rect.width, rect.height);
                                try {
                                    dirtyComponent.paint(g);
                                } finally {
                                    g.dispose();
                                }
                            }
                        }
```

So when my JLabel was opaque: we entered the first branch (where `dirtyComponent` is the JLabel). Here Swing tightly controls this process, and eventually we call `c.paintToOffscreen`, where c is the JLabel and the Graphics2D is a reference to RepaintManager's offscreen image. Once the JComponent is finished, then the PaintManager calls:
```
                            final Graphics2D g2d = (Graphics2D) g;
                            final Composite oldComposite = g2d.getComposite();
                            g2d.setComposite(AlphaComposite.Src);
                            g2d.drawImage(image, x, y, c);
                            g2d.setComposite(oldComposite);
```
This time `g` is the Window's Graphics2D that relates to the actual surface data on the monitor.

Specifically the stacktrace I'm seeing in JDK 19 when I repainted my opaque JLabel is:
```
paintDoubleBufferedImpl:1663, RepaintManager$PaintManager (javax.swing)
paintDoubleBuffered:1631, RepaintManager$PaintManager (javax.swing)
paint:1569, RepaintManager$PaintManager (javax.swing)
paint:1336, RepaintManager (javax.swing)
_paintImmediately:5266, JComponent (javax.swing)
paintImmediately:5076, JComponent (javax.swing)
run:878, RepaintManager$4 (javax.swing)
run:861, RepaintManager$4 (javax.swing)
executePrivileged:776, AccessController (java.security)
doPrivileged:399, AccessController (java.security)
doIntersectionPrivilege:86, ProtectionDomain$JavaSecurityAccessImpl (java.security)
paintDirtyRegions:861, RepaintManager (javax.swing)
paintDirtyRegions:834, RepaintManager (javax.swing)
prePaintDirtyRegions:784, RepaintManager (javax.swing)
run:1897, RepaintManager$ProcessingRunnable (javax.swing)
dispatch$$$capture:318, InvocationEvent (java.awt.event)
dispatch:-1, InvocationEvent (java.awt.event)
```

>  2. Why the double-buffer in the RepaintManager does not handle this? If no buffers are used, then probably we should have one to render everything to it and then blit to Window?

In the original unit test: the JLabel is not opaque, so the `dirtyComponent` we set out to repaint was our root JWindow. So we entered the second branch in the code above and called `dirtyComponent.paint(g)`.

Once we entered Swing rendering code: all our rendering really was double-buffered to the appropriate offscreen image. But the problem is Window.java has that one call to `fillRect` that is outside of Swing's rendering model. (So in a way this is a subtle clash of Swing vs AWT code.)

mickleness commented:
So I'm generally hearing 3 responses to this ticket (so far):

1. My interpretation: we have this one known problem (where Window.java is calling `fillRect(..)`), and if we just flip a switch and avoid that code then the flicker is gone.
2. Alan's interpretation (see [8209329](https://bugs.openjdk.org/browse/JDK-8209329) ): we can eliminate the flicker by better synchronizing when the Window's surface data is flushed. (So `Window.paint(..)` can do anything it wants to: if we control when the monitor refreshes there won't be a flicker.)
3. Sergey's interpretation: if we make sure the Graphics we pass to `dirtyComponent.paint(..)` is the offscreen buffer, then that should also resolve the flicker.

Does this sound like a fair assessment, and if so: how do we decide which response to pursue?

mrserb commented:
improving synchronization will make the same effect as using the double-buffer for this translucent windows, but probably adding the  double-buffer is simpler to implement. You can also check how the "swing.bufferPerWindow" property is implemented, we can try to improve it for this issue and set by default.

mrserb commented:
For the possible "synchronization" change, note that on macOS the rendering to the Window is always done via backbuffer which then blits to the CALayer. We can try to delay the blit until the window is fully rendered.

DamonGuy commented on src/java.desktop/share/classes/java/awt/Window.java:
```diff
@@ -3949,7 +3950,7 @@ public void paint(Graphics g) {
                 if (gg instanceof Graphics2D) {
                     gg.setColor(getBackground());
-                    ((Graphics2D)gg).setComposite(AlphaComposite.getInstance(AlphaComposite.SRC));
+                    ((Graphics2D) gg).setComposite(AlphaComposite.getInstance(AlphaComposite.SRC));
```
Why was the import for SunHints added if the only edit to this file was adding a space?

DamonGuy replied:
Nevermind I saw this wrong when looking through the commit history.

mickleness replied:
Thanks for checking, though. I just pushed a change to roll back this diffs. https://github.com/openjdk/jdk/pull/12993/commits/f369ccde6d73537948d6f5007ae5d0a2b847d8ac
